Web extension resources return incorrect MIME type

amvanbaren opened the pull request:
Contributes to #468
Use Apache Tika to detect mimetype

jeanp413 commented:
@amvanbaren just found that mime types are listed in the `[Content_Types].xml` file inside the vsix :facepalm: So we just need to parse it and use the rules in there
```xml
<?xml version="1.0" encoding="utf-8"?>
<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
	<Default Extension=".json" ContentType="application/json"/>
	<Default Extension=".vsixmanifest" ContentType="text/xml"/>
	<Default Extension=".png" ContentType="image/png"/>
	<Default Extension=".md" ContentType="text/markdown"/>
	<Default Extension=".js" ContentType="application/javascript"/>
	<Default Extension=".map" ContentType="application/json"/>
	<Default Extension=".txt" ContentType="text/plain"/>
	<Default Extension=".woff2" ContentType="font/woff2"/>
	<Default Extension=".html" ContentType="text/html"/>
	<Default Extension=".css" ContentType="text/css"/>
</Types>
```

jeanp413 commented:
Some more info https://docs.microsoft.com/en-us/visualstudio/extensibility/the-structure-of-the-content-types-dot-xml-file?view=vs-2022#attribute-name-attribute-1
vsix files should be `application/zip` and all other file types `application/octet-stream`

amvanbaren commented:
> I just found that mime types are listed in the [Content_Types].xml file inside the vsix 🤦 So we just need to parse it and use the rules in there.

Ok, I'll use that file to assign the right content type to a file resource. If the resource isn't listed then `application/octet-stream` is used.

>  vsix files should be application/zip and all other file types application/octet-stream

When I download a vsix file, then the content-type is `application/vsix`
![content-type-application-vsix](https://user-images.githubusercontent.com/14129099/173017123-e1b8a223-7e60-42ba-9b79-a00cec89145d.png)

jeanp413 commented:
> When I download a vsix file, then the content-type is application/vsix

Then their docs are outdated :sweat_smile:

jeanp413 commented on server/src/test/java/org/eclipse/openvsx/adapter/VSCodeAdapterTest.java:
```diff
@@ -246,6 +247,15 @@ public void testAssetNotFound() throws Exception {
+        mockMvc.perform(get("/vscode/asset/{namespace}/{extensionName}/{version}/{assetType}",
+                "redhat", "vscode-yaml", "0.5.2", "Microsoft.VisualStudio.Services.VSIXPackage"))
+                .andExpect(status().isOk())
+                .andExpect(header().string(HttpHeaders.CONTENT_TYPE, "application/zip"));
```
Shouldn't it be `"application/vsix"`?

jeanp413 replied:
Just checked it in MS marketplace, in this case is `application/zip` :thinking:

amvanbaren replied:
Ok, let's use `application/zip`

jeanp413 commented:
Need to test what happens when some files don't have file extension, e.g sometimes `README` files don't have the `.md` extension, maybe we should add some default rules by filename

filiptronicek commented:
@jeanp413 if there is no extension present, it will not be in the `[Content_Types].xml` file. Do you think there could be cases in which serving everything with no extension with a `text/plain` would not work? 

This is just concerning web extensions, correct? So there will be no binaries served, which are the only thing in my mind that don't usually have extensions.

cc @amvanbaren

jeanp413 commented:
Thanks for testing this, @filiptronicek, from the MS docs anything else should be `application/octet-stream`.
Looking at vsce source code they allow [license without extension](https://github.dev/microsoft/vscode-vsce/blob/75456f2e56fa3911a998f37f72215ec3d26e88e3/src/package.ts#L881-L893) but then they add the file extension if [it's missing when packaged](https://github.dev/microsoft/vscode-vsce/blob/75456f2e56fa3911a998f37f72215ec3d26e88e3/src/package.ts#L900-L904) so we won't have any issue :tada:

jeanp413 commented on server/src/test/java/org/eclipse/openvsx/adapter/VSCodeAdapterTest.java:
```diff
@@ -542,45 +595,87 @@ private ExtensionVersion mockExtension(String targetPlatform) throws JsonProcess
+        cssWebResourceFile.setContentType("text/css");
+        Mockito.when(repositories.findFileByTypeAndName(extVersion, FileResource.RESOURCE, "extension/public/static/css/main.css"))
+                .thenReturn(cssWebResourceFile);
+        var cssChunkWebResourceFile = new FileResource();
+        cssChunkWebResourceFile.setExtension(extVersion);
+        cssChunkWebResourceFile.setName("extension/public/static/css/main.9cab4879.chunk.css");
+        cssChunkWebResourceFile.setType(FileResource.RESOURCE);
+        cssChunkWebResourceFile.setStorageType(STORAGE_DB);
+        cssChunkWebResourceFile.setContent(".root { margin: 0 auto; }".getBytes());
+        cssChunkWebResourceFile.setContentType("text/css");
+        Mockito.when(repositories.findFileByTypeAndName(extVersion, FileResource.RESOURCE, "extension/public/static/css/main.9cab4879.chunk.css"))
+                .thenReturn(cssChunkWebResourceFile);
```
Let's add one more test for the default mime type `application/octet-stream`

jeanp413 commented:
@amvanbaren I saw openvsx got finally deployed :tada:, can we merge this so it's shipped in next deployment :slightly_smiling_face: 
BTW, can you share until which commit got deployed?

amvanbaren commented:
> can we merge this so it's shipped in next deployment 🙂

Yes, I'll rebase and complete this PR.

> BTW, can you share until which commit got deployed?

Up until PR #517 is deployed. Production is 25 commits behind `master`.

amvanbaren commented:
> @amvanbaren let's merge 🙏

@jeanp413 Has this been reviewed and tested?

jeanp413 commented:
I thought it was already approved :sweat_smile: , will take a look

amvanbaren commented:
This PR is missing a migration: https://github.com/eclipse/openvsx/issues/468#issuecomment-1323388969

jeanp413 commented:
@filiptronicek if you have some free time could you test this :pray: ?

filiptronicek reviewed (commented):
What seems to have happened is that now everything is an octet stream, even though the files' extensions are listed in the content types file along with their MIME types. 

<img width="1242" alt="image" src="https://user-images.githubusercontent.com/29888641/203849297-96c6b6b5-a6f6-4656-8eef-c288c3cd0ac8.png">
<img width="1262" alt="image" src="https://user-images.githubusercontent.com/29888641/203849794-74cd4c55-fd40-4416-ac70-c46daff86a1d.png">

kineticsquid commented:
Same priority as #468

Commit:
Web extension resources return incorrect MIME type

Use [Content_Types].xml to set FileResource contentType
Add test case for default content type: application/octet-stream
Add migration to set FileResource.contentType
Set contentType for RESOURCE type
Remove dot ('.') if extension starts with a dot
Use utf-8 charset for text resources.

amvanbaren commented:
Ok, I've fixed the PR. @filiptronicek Do you have time to test it?
